### DISCUSSION TITLE
EVP_PKEY_keygen takes unusually long sometimes

### DISCUSSION THREAD

**Discussion by agneshSE** [openssl/openssl#24174] [Q&A]
I am currently using **openSSL 3.0.8** and am facing random extended delay in the key generation function, EVP_PKEY_keygen. Below are the logs and code snippet of the function I used for testing.

![testFunc 1](https://github.com/openssl/openssl/assets/159979681/1cbaa1f4-6732-480e-a364-06a84b26fdac)
![testFunc](https://github.com/openssl/openssl/assets/159979681/6ff75d67-20a3-41c9-8ebf-aabe8d4c3a48)
![image](https://github.com/openssl/openssl/assets/159979681/0c06026e-8843-48da-a60c-5d756027ab31)

**Comment by t8m** [comment]
It would be interesting to know if you see any periods of time during the key generation when the process is in waiting state - i.e. CPU is not used. Otherwise this looks like a normal variation of time based on whether the keygen is lucky to find suitable primes earlier or later.

**Reply by agneshSE** [reply]
Hi Tomáš, Thanks for the reply.
the process for the entire duration does not go into and waiting state, also this extended delay was not observed in 1.1.x before moving to 3.0.8. 

Was there any changes in the function implementation? I would also like to know if updating to version 3.1 help reduce this time delay?

**Reply by t8m** [reply]
The generation algorithm was changed in 3.0. There were some speed ups added later. Not sure if 3.1 reduces the keygen time significantly though. It might be the bigger speedups came in 3.2.

**Comment by slontis** [comment]
RSA keygen is using a slightly more complex algorithm than it was in 1.1.x
See PR #19578 and PR #19579 (Basically calling a function many times that was changed to be constant time at some point resulted in a slowdown - all the RSA keygen algorithms do the same kind of tests, so the chosen algorithm is not the bottleneck).
Unfortunately these changes were deemed as performance fixes, not bug fixes so they were not backported to 3.0 after 3.0 became a LTS release.
